File: LeetCode/Sort List.cpp
#include <iostream>
using namespace std;
/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
    ListNode *sortList(ListNode *head, int len) {
        if (len <= 1)
            return head;

        ListNode *curr = head;
        for (int i = 0; i < len / 2; i++) {
            curr = curr->next;
        }

        ListNode *left = sortList(head, len / 2);
        ListNode *right = sortList(curr, len - len / 2);

        head = curr = nullptr;
        int i, j;
        for (i = j = 0; i < len / 2 && j < len - len / 2; ) {
            ListNode *tmp = nullptr;
            if (left->val < right->val) {
                tmp = left;
                left = left->next;
                i++;
            } else {
                tmp = right;
                right = right->next;
                j++;
            }

            if (curr) curr->next = tmp;
            curr = tmp;
            if (head == nullptr) head = curr;
        }

        if (i < len / 2)
            curr->next = left;
        if (j < len - len / 2)
            curr->next = right;

        return head;
    }

public:
    ListNode* sortList(ListNode* head) {
        ListNode *curr;

        int len = 0;
        for (curr = head; curr; curr = curr->next)
            len++;

        head = sortList(head, len);

        curr = head;
        for (int i = 0; i < len - 1; i++)
            curr = curr->next;

        if (curr) curr->next = nullptr;
        return head;
    }
};

int main(int argc, char *argv[])
{
    Solution sol;
    ListNode n1(1);
    ListNode n2(2);
    ListNode n3(3);
    ListNode n4(4);

    n2.next = &n1;
    n1.next = &n4;
    n4.next = &n3;

    ListNode *n = sol.sortList(nullptr);

    for (; n; n = n->next)
        cout << n->val << endl;

    return 0;
}

File: LeetCode_Java/src/SortList.java
public class SortList {
    /**
     * Definition for singly-linked list.
     * public class ListNode {
     *     int val;
     *     ListNode next;
     *     ListNode(int x) { val = x; }
     * }
     */

    private ListNode mergeList(ListNode leftList, ListNode rightList) {
        ListNode head = null;
        ListNode curNode = null;

        while (leftList != null && rightList != null) {
            ListNode nextNode = null;
            if (leftList.val < rightList.val) {
                nextNode = leftList;
                leftList = leftList.next;
            } else {
                nextNode = rightList;
                rightList = rightList.next;
            }

            if (head == null) {
                head = nextNode;
                curNode = head;
            }
            curNode.next = nextNode;
            curNode = nextNode;
        }
        if (leftList != null) {
            curNode.next = leftList;
        } else if (rightList != null) {
            curNode.next = rightList;
        }

        return head;
    }

    private ListNode mergeSort(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        System.out.println(slow.val);
        ListNode rightList = mergeSort(slow.next);
        slow.next = null;
        ListNode leftList = mergeSort(head);

        return mergeList(rightList, leftList);
    }

    public ListNode sortList(ListNode head) {
        return mergeSort(head);
    }
}
